lib/rsync_store.ml

```diff
@@ -0,0 +1,153 @@
+
+  let delete dir =
+    (* Sudo and rm... yikes *)
+    Os.sudo [ "rm"; "-r"; dir ]
+
+  let rsync = [ "rsync"; "-aq" ]
+
+  let copy_children ~src ~dst =
+    let cmd = rsync @ [ Fmt.str "%s/" src; dst ] in
+    Os.ensure_dir dst;
+    Os.sudo cmd
+end
+
+module Path = struct
+  let state_dirname = "state"
+  let cache_dirname = "cache"
+  let cache_tmp_dirname = "cache-tmp"
+
+  let result_dirname = "result"
+  let result_tmp_dirname = "result-tmp"
+
+  let dirs root =
+    List.map ((/) root)
+    [ state_dirname; cache_dirname; cache_tmp_dirname; result_dirname; result_tmp_dirname ]
+
+  let result t id = t.path / result_dirname / id
+  let cache t id = t.path / cache_dirname / id
+
+  let cache_tmp t n id = t.path / cache_tmp_dirname / Fmt.str "%i-%s" n id
+
+  let result_tmp t id = t.path / result_tmp_dirname / id
+end
+
+let create ~path =
+  Rsync.create path >>= fun () ->
+  Lwt_list.iter_s Rsync.create (Path.dirs path) >|= fun () ->
+  { path; caches = Hashtbl.create 10; next = 0 }
+
+let build t ?base ~id fn =
+  Log.debug (fun f -> f "rsync: build %S" id);
+  let result = Path.result t id in
+  let result_tmp = Path.result_tmp t id in
+  begin match base with
+  | None -> Rsync.create result_tmp
+  | Some base -> Rsync.copy_children ~src:(Path.result t base) ~dst:result_tmp
+  end
+  >>= fun () ->
+  Lwt.try_bind
+      (fun () -> fn result_tmp)
+      (fun r ->
+      begin match r with
+          | Ok () -> Rsync.copy_children ~src:result_tmp ~dst:result
+          | Error _ -> Lwt.return_unit
+      end >>= fun () ->
+      Rsync.delete result_tmp >>= fun () ->
+      Lwt.return r
+      )
+  (fun ex ->
+      Log.warn (fun f -> f "Uncaught exception from %S build function: %a" id Fmt.exn ex);
+      Rsync.delete result_tmp >>= fun () ->
+      Lwt.fail ex
+  )
+
+let delete t id =
+  let path = Path.result t id in
+  match Os.check_dir path with
+    | `Present -> Rsync.delete path
+    | `Missing -> Lwt.return_unit
+
+let result t id =
+  let dir = Path.result t id in
+  match Os.check_dir dir with
+  | `Present -> Some dir
+  | `Missing -> None
+
+let state_dir t = t.path / Path.state_dirname
+
+let get_cache t name =
+  match Hashtbl.find_opt t.caches name with
+  | Some c -> c
+  | None ->
+    let c = { lock = Lwt_mutex.create (); gen = 0 } in
+    Hashtbl.add t.caches name c;
+    c
+
+let cache ~user t name =
+  let cache = get_cache t name in
+  Lwt_mutex.with_lock cache.lock @@ fun () ->
+  let tmp = Path.cache_tmp t t.next name in
+  t.next <- t.next + 1;
+  let snapshot = Path.cache t name in
+  (* Create cache if it doesn't already exist. *)
+  begin match Os.check_dir snapshot with
+      | `Missing -> Rsync.create snapshot
+      | `Present -> Lwt.return_unit
+  end >>= fun () ->
+  (* Create writeable clone. *)
+  let gen = cache.gen in
+  Rsync.copy_children ~src:snapshot ~dst:tmp >>= fun () ->
+  let { Obuilder_spec.uid; gid } = user in
+  Os.sudo ["chown"; Printf.sprintf "%d:%d" uid gid; tmp] >>= fun () ->
```

Comment:
This change went into #88 but I've just realised that macOS ships with a much older version of `rsync` that doesn't support `--chown`. There are two options:
 - Go back to this implementation so Obuilder will work with the default rsync macOS comes with.
 - Force people to download a newer version -- note this is not necessarily super straightforward because `brew install rsync` will put it in `/usr/local` and won't be available when `obuilderfs` is active, so the user will probably have to build from source and put the binary and libs elsewhere.

Any thoughts ?

Reply:
Yeah, I guess we can revert that then. I'm not used to things being older than debian/stable! (and, indeed, older than debian/oldstable too!!)

Might be a performance hit, though. I remember that running `sudo chown` in Dockerfiles is massively slower than setting the ownership in the `COPY`.